server/src/main/java/org/uiautomation/ios/wkrdp/WebKitNotificationListener.java

```diff
@@ -27,162 +35,278 @@
+      handle(applicationSentListingMessage);
+    } else if (message instanceof ApplicationDataMessage) {
+      // ApplicationDataMessage m = (ApplicationDataMessage)message;
+    } else if (message instanceof ApplicationConnectedMessage) {
+      ApplicationConnectedMessage applicationConnectedMessage = ApplicationConnectedMessage.class.cast(message);
+      handle(applicationConnectedMessage);
     }
+  }
+
+  private void handle(ReportSetupMessage reportSetupMessage) {
+    driver.setDevice(reportSetupMessage.getDevice());
+    sync.signalSimRegistered();
+  }
 
-    if (message instanceof ReportConnectedApplicationsMessage) {
-      ReportConnectedApplicationsMessage m = (ReportConnectedApplicationsMessage) message;
-      if (m.getApplications().size() == 0) {
-        log.warning("ReportConnectedApplicationsMessage reported 0 app.");
+  private void handle(ReportConnectedApplicationsMessage reportConnectedApplicationsMessage) {
+    try {
+      applicationRegistrationLock.lock();
+      registeredApplications.addAll(reportConnectedApplicationsMessage.getApplications());
+      if (!isDriverSuccessfullyNotified()) {
+        log.log(Level.WARNING,
+            "No connectable applications found in ReportConnectedApplicationsMessage, will wait before signalling driver");
+      }
+    } finally {
+      applicationRegistrationLock.unlock();
+    }
+  }
+
+  private void handle(ApplicationSentListingMessage applicationSentListingMessage) {
+    if (!isValidForProcessing(applicationSentListingMessage)) {
+      return;
+    }
+    try {
+      pagesProcessLock.lock();
+      boolean pagesSimilar = arePagesNotChanged(applicationSentListingMessage);
+      if (log.isLoggable(Level.FINE)) {
+        log.fine("pages " + (pagesSimilar ? "are equal " : "have changed ") + ": " + driver.getPages() + "-> "
+            + applicationSentListingMessage.getPages() + ": " + applicationSentListingMessage);
+      }
+      if (pagesSimilar) {
+        driver.setPages(applicationSentListingMessage.getPages()); // Update the titles.
+        return;
+      }
+      if (session == null || session.getApplication().isSafari()) {
+        processPagesForSafariApp(applicationSentListingMessage);
       } else {
-        driver.setApplications(m.getApplications());
+        processPagesForHybridApp(applicationSentListingMessage);
+      }
+    } finally {
+      pagesProcessLock.unlock();
+    }
+  }
+
+  private void handle(ApplicationConnectedMessage applicationConnectedMessage) {
+    try {
+      applicationRegistrationLock.lock();
+      registeredApplications.add(applicationConnectedMessage.getApplication());
+      if (!isDriverSuccessfullyNotified()) {
+        log.log(Level.WARNING,
+            "No connectable applications found in ApplicationConnectedMessage, will wait before signalling driver");
+      }
+    } finally {
+      applicationRegistrationLock.unlock();
+    }
+  }
+
+  private boolean isDriverSuccessfullyNotified() {
```

Comment:
The 2 callers of this method check for a `false` return type and log the same message when the condition occurs. What about logging the message in this method instead and re-factoring a bit to accomplish the same net result?